hyperspy/roi.py

```diff
@@ -1511,6 +1514,509 @@ def __call__(self, signal, out=None, axes=None, order=0):
             out.events.data_changed.trigger(out)
 
 
+class PolygonROI(BaseInteractiveROI):
+    """Selects a polygonal region in a 2D space. The coordinates of the
+    polygon vertices are given in the `vertices` attribute, which is a
+    list where each entry is a tuple (x, y) of a vertex's coordinates.
+    An edge runs from the final to the first point in the list.
+    If a polygon self overlaps, the overlapping areas may be considered
+    outside of the polygon and masked away.
+    """
+
+    vertices = []
+    _ndim = 2
+
+    def __init__(self, vertices=None):
+        """
+        Parameters
+        ----------
+        vertices : list of tuples
+            List containing (x, y) values of the vertices of a polygon."""
+        super().__init__()
+
+        if vertices:
+            self.set_vertices(vertices)
+
+    @property
+    def parameters(self):
+        return {"vertices": self.vertices}
+
+    def __getitem__(self, *args, **kwargs):
+        _tuple = tuple(self.vertices)
+        return _tuple.__getitem__(*args, **kwargs)
+
+    def __repr__(self):
+        para = []
+        for name, value in self.parameters.items():
+            parastr = f"{name}=["
+            parastr += ", ".join(f"({x:G}, {y:G})" for x, y in value)
+            parastr += "]"
+            para.append(parastr)
+        return f"{self.__class__.__name__}({', '.join(para)})"
+
+    def is_valid(self):
+        """
+        The polygon is defined as valid if either zero or more than two
+        vertices are fully defined.
+        """
+        return (
+            len(self.vertices) == 0
+            or len(self.vertices) > 2
+            and all(
+                (None not in vertex and len(vertex) == 2) for vertex in self.vertices
+            )
+        )
+
+    def set_vertices(self, vertices):
+        """Sets the vertices of the polygon to the `vertices` argument,
+        where each vertex is to be given as a tuple `(x, y)` where `x`
+        and `y` are its coordinates. The list is set to loop around,
+        such that an edge runs from the final to the first vertex in
+        the list.
+
+        Parameters
+        ----------
+        vertices : list of tuples
+            List of (x, y) values of the vertices of the polygon.
+        """
+        old_vertices = self.vertices
+        self.vertices = vertices
+        if not self.is_valid():
+            self.vertices = old_vertices
+            raise ValueError(
+                f"`vertices` is not an empty list or a list of fully defined two-dimensional \
+                    points with at least three entries:\n{vertices}"
+            )
+        if self.widgets:
+            self._update_widgets()
+
+    def _apply_roi(self, signal, inverted=False, out=None, axes=None, other_rois=None):
```

Comment:
The reason for including the other ROIs here is that it's more efficient to rasterize all of them together and then slice the signal once. But some other ways of doing this could be:
- In `combine_roi`, slice the signal once for each `PolygonROI` and stitch them together. Should be fairly straightforward but maybe a bit inefficient.
- Let `_apply_roi` take in a boolean mask it can combine with the polygon, which can be created with `hs.roi.create_mask_from_rois`. Then it only needs to slice once without needing the other ROIs directly. Would then also combine with other types of ROIs.
I think the latter would be best due to efficiency and also it would be easier to coordinate the parameters, such as x- and y-limits for cropping the finished signal. I will take a look a bit later on.